pathd/path_pcep_controller.c

```diff
@@ -779,7 +929,161 @@ int pcep_main_event_handler(struct thread *thread)
+			break;
+		}
+	}
+	return previous_best_pce != -1 ? pcc[previous_best_pce]->id : 0;
+}
+int calculate_best_pce(struct ctrl_state *ctrl_state)
+{
+	int best_precedence = 255; // DEFAULT_PCE_PRECEDENCE;
+	int best_pce = -1;
+	int one_connected_pce = -1;
+	int previous_best_pce = -1;
+	int step_0_best = -1;
+	int step_0_previous = -1;
+
+	if (!ctrl_state || !ctrl_state->pcc_count)
+		return 0;
+
+	struct pcc_state **pcc = &ctrl_state->pcc[0];
+
+	// Get state
+	for (int i = 0; i < MAX_PCC; i++) {
+		if (pcc[i] && pcc[i]->pce_opts) {
+			if (pcc[i]->is_best == true) {
+				step_0_best = i;
+			}
+			if (pcc[i]->previous_best == true) {
+				step_0_previous = i;
+			}
+		}
+	}
+	for (int i = 0; i < MAX_PCC; i++) {
+		if (pcc[i] && pcc[i]->pce_opts) {
+			zlog_debug(
+				"calculate all : i (%i) is_best (%i) previous_best (%i)   ",
+				i, pcc[i]->is_best, pcc[i]->previous_best);
+		}
+	}
+
+	// Calculate best
+	for (int i = 0; i < MAX_PCC; i++) {
+		if (pcc[i] && pcc[i]->pce_opts
+		    && pcc[i]->status != PCEP_PCC_DISCONNECTED) {
+			one_connected_pce = i; // In case none better
+			if (pcc[i]->pce_opts->precedence <= best_precedence) {
+				if (best_pce != -1
+				    && pcc[best_pce]->pce_opts->precedence
+					       == pcc[i]->pce_opts->precedence
+				    && ipaddr_cmp(
+					       &pcc[i]->pce_opts->addr,
+					       &pcc[best_pce]->pce_opts->addr)
+					       > 0) {
+					// collide of precedences so compare ip
+					best_pce = i;
+				} else {
+					if (!pcc[i]->previous_best) {
+						best_precedence =
+							pcc[i]->pce_opts
+								->precedence;
+						best_pce = i;
+					}
+				}
+			}
+		}
+	}
 
+	zlog_debug("calculate data : sb (%i) sp (%i) oc (%i) b (%i)  ",
+		   step_0_best, step_0_previous, one_connected_pce, best_pce);
+
+	// Changed of state so ...
+	if (step_0_best != best_pce) {
+		// Calculate previous
+		previous_best_pce = step_0_best;
+		// Clean state
+		if (step_0_best != -1) {
+			pcc[step_0_best]->is_best = false;
+		}
+		if (step_0_previous != -1) {
+			pcc[step_0_previous]->previous_best = false;
+		}
+
+		// Set previous
+		if (previous_best_pce != -1) {
+			pcc[previous_best_pce]->previous_best = true;
+			zlog_debug("previous best pce (%i) ",
+				   previous_best_pce + 1);
+		}
+
+
+		// Set best
+		if (best_pce != -1) {
+			pcc[best_pce]->is_best = true;
+			zlog_debug("best pce (%i) ", best_pce + 1);
+		} else {
+			if (one_connected_pce != -1) {
+				best_pce = one_connected_pce;
+				pcc[one_connected_pce]->is_best = true;
+				zlog_debug(
+					"one connected best pce (default) (%i) ",
+					one_connected_pce + 1);
+			} else {
+				for (int i = 0; i < MAX_PCC; i++) {
+					if (pcc[i] && pcc[i]->pce_opts) {
+						best_pce = i;
+						pcc[i]->is_best = true;
+						zlog_debug(
+							"(disconnected) best pce (default) (%i) ",
+							i + 1);
+						break;
+					}
+				}
+			}
+		}
+	}
+
+
+	return best_pce == -1 ? 0 : pcc[best_pce]->id;
+}
+int pcep_ctrl_get_free_pcc_idx(struct pcc_state **pcc_state)
```

Comment:
Why not get the controller state as parameter, this doesn't fell safe to assume the size of what you get as a parameter is MAX_PCC, you can only assume that from the array in the controller state.

Reply:
changed